File: java/RomanToInteger.java
public class RomanToInteger {
    public int romanToInt(String s) {
        int num = 0;
        if (s == null || s.length() == 0) {
            return num;
        }

        int len = s.length();
        int[] values = new int[128];
        values['I'] = 1;
        values['V'] = 5;
        values['X'] = 10;
        values['L'] = 50;
        values['C'] = 100;
        values['D'] = 500;
        values['M'] = 1000;

        for (int i = 0; i < len - 1; ++i) {
            char v = s.charAt(i);
            char w = s.charAt(i + 1);
            if (values[v] < values[w]) {
                num -= values[v];
            } else {
                num += values[v];
            }
        }
        num += values[s.charAt(len - 1)];
        return num;
    }

    public static void main(String[] args) {
        RomanToInteger solution = new RomanToInteger();
        System.out.println(solution.romanToInt("MDCVI"));
        System.out.println(solution.romanToInt("XLIX"));
    }
}

File: cc/roman_to_integer.cc
#include "base_include.h"

class RomanToInteger {
public:
    int romanToInt(string s) {
        int hash[128];
        hash['M'] = 1000;
        hash['D'] = 500;
        hash['C'] = 100;
        hash['L'] = 50;
        hash['X'] = 10;
        hash['V'] = 5;
        hash['I'] = 1;
        int num = 0;
        int n = s.size();
        for (int i = 0; i < n - 1; ++i) {
            if (hash[(int) s[i]] < hash[(int) s[i + 1]]) {
                num -= hash[(int) s[i]];
            } else {
                num += hash[(int) s[i]];
            }
        }
        num += hash[(int) s[n - 1]];
        return num;
    }
};

int main(int argc, char *argv[]) {
    return 0;
}
